vltt/assignment3/solution.cpp

```diff
@@ -0,0 +1,197 @@
+#define _SILENCE_TR1_NAMESPACE_DEPRECATION_WARNING
+#include <iostream>
+#include <algorithm>
+#include <iomanip>
+#include <vector>
+#include <map>
+#include <unordered_set>
+#include <set>
+#include <string>
+#include <gtest/gtest.h>
+
+#define EXPECT_EQ(expected, actual)                                       \
+  if ((expected) != (actual)) {                                           \
+    std::cout << __FILE__ << ":" << __LINE__ << " Failed: "               \
+              << "Statement (" << #actual << ") is " << actual            \
+              << ", expected " << (expected) << " (" << #expected << ")." \
+              << std::endl;                                               \
+  } else {                                                                \
+    std::cout << __FILE__ << ":" << __LINE__ << " True " << std::endl;  \
+  }
+  
+
+template <class T>
+inline std::ostream& operator<<(std::ostream& os, const std::set<T>& s) {
+    os << "{";
+    for (auto &it : s) {
+        os << it;
+        std::cout << ", " ;
+    }
+    os << "}";
+    return os;
+}
+
+class Dictionary {
+private:
+    class Trie {
+    private:
+        class Node {
+        private:
+            std::map<char, Node*> go;
+        public:
+            bool terminal = false;
+            void add(const char &letter) {
+                go[letter] = new Node;
+            }
+            bool isHere(const char &letter) {
+                return go.find(letter) != go.end();
+            }
+            Node *next(const char &letter) {
+                if (go.find(letter) == go.end()) return nullptr;
+                return go[letter];
+            }
+        };
+        Node *start;
+    public:
+        Trie() {
+            start = new Node;
+        }
+        //add new word to the Trie
+        void add(const std::string &word) {
+            if (word.empty()) return;
+            Node *cur = start;
+            for (int i = 0; i < word.length(); i++) {
+                char letter = word[i];
+                if (cur->isHere(letter) == false) {
```

Comment:
[optional]: a slightly more readable way would be:

`if (!cur->isHere(letter))`